spec/index/categorical_index_spec.rb

```diff
@@ -0,0 +1,89 @@
+require 'spec_helper.rb'
+
+describe Daru::CategoricalIndex do
+  context "#pos" do
+    context "when the category is non-numeric" do
+      let(:idx) { described_class.new [:a, :b, :a, :a, :c] }
+
+      context "single category" do
+        subject { idx.pos :a }
+        
+        it { is_expected.to eq [0, 2, 3] }
+      end
+      
+      context "multiple categories" do
+        subject { idx.pos :a, :c }
+        
+        it { is_expected.to eq [0, 2, 3, 4] }
+      end
+
+      context "invalid category" do
+        it { expect { idx.pos :e }.to raise_error IndexError }
+      end
+
+      context "positional index" do
+        it { expect(idx.pos 0).to eq 0 }
+      end
+
+      context "invalid positional index" do
+        it { expect { idx.pos 5 }.to raise_error IndexError }
+      end
+
+      context "multiple positional indexes" do
+        subject { idx.pos 0, 1, 2 }
+
+        it { is_expected.to be_a Array }
+        its(:size) { is_expected.to eq 3 }
+        it { is_expected.to eq [0, 1, 2] }
+      end
+    end
+
+    context "when the category is numeric" do
+      let(:idx) { described_class.new [0, 1, 0, 0, 2] }
+
+      context "first preference to category" do
+        subject { idx.pos 0 }
+
+        it { is_expected.to be_a Array }
+        its(:size) { is_expected.to eq 3 }
+        it { is_expected.to eq [0, 2, 3] }
+      end
+
+      context "second preference to positional index" do
+        subject { idx.pos 3 }
+
+        it { is_expected.to eq 3 }
+      end
+    end
+  end
+  
+  context "#subset" do
+    let(:idx) { described_class.new [:a, 1, :a, 1, :c] }
+    
+    context "single index" do
+      context "multiple instances" do
+        subject { idx.subset :a }
+        
+        it { is_expected.to be_a described_class }
+        its(:size) { is_expected.to eq 2 }
+        its(:to_a) { is_expected.to eq [:a, :a] }
+      end
+    end
+    
+    context "multiple indexes" do
+      subject { idx.subset :a, 1 }
+      
+      it { is_expected.to be_a described_class }
+      its(:size) { is_expected.to eq 4 }
+      its(:to_a) { is_expected.to eq [:a, :a, 1, 1] }
+    end
```

Comment:
According to the definition of `subset` this seems to be correct. But I also think that this behavior is rather unintuitive. If I take a subset of my data, I usually don't expect the observations to be permuted in the result. (Especially dangerous if for example you have two datasets for the same subjects in the same order, and after some data preprocessing the order of the subjects in the two datasets suddenly doesn't quite match up anymore. This actually happened to me once, and I found the mistake only weeks later after the paper was already submitted, because the results still looked reasonable even though they were totally wrong.)
@lokeshh Is there a reason why it is better to permute the order in this way? Or am I missing something else here?